rotkehlchen/chain/evm/decoding/compound/v3/decoder.py

```diff
@@ -0,0 +1,267 @@
+import logging
+from collections.abc import Callable
+from typing import TYPE_CHECKING, Any
+
+from rotkehlchen.assets.asset import EvmToken
+from rotkehlchen.assets.utils import get_or_create_evm_token
+from rotkehlchen.chain.ethereum.utils import asset_normalized_value
+from rotkehlchen.chain.evm.decoding.interfaces import DecoderInterface
+from rotkehlchen.chain.evm.decoding.structures import (
+    DEFAULT_DECODING_OUTPUT,
+    ActionItem,
+    DecoderContext,
+    DecodingOutput,
+)
+from rotkehlchen.chain.evm.decoding.types import CounterpartyDetails
+from rotkehlchen.constants.resolver import evm_address_to_identifier
+from rotkehlchen.errors.asset import UnknownAsset, WrongAssetType
+from rotkehlchen.globaldb.handler import GlobalDBHandler
+from rotkehlchen.history.events.structures.types import HistoryEventSubType, HistoryEventType
+from rotkehlchen.logging import RotkehlchenLogsAdapter
+from rotkehlchen.types import ChecksumEvmAddress, EvmTokenKind, EvmTransaction
+from rotkehlchen.utils.misc import hex_or_bytes_to_address, hex_or_bytes_to_int
+
+from .constants import (
+    COMPOUND_V3_SUPPLY,
+    COMPOUND_V3_SUPPLY_COLLATERAL,
+    COMPOUND_V3_WITHDRAW,
+    COMPOUND_V3_WITHDRAW_COLLATERAL,
+    CPT_COMPOUND_V3,
+    REWARD_CLAIMED,
+)
+
+if TYPE_CHECKING:
+    from rotkehlchen.chain.evm.decoding.base import BaseDecoderTools
+    from rotkehlchen.chain.evm.node_inquirer import EvmNodeInquirer
+    from rotkehlchen.chain.evm.structures import EvmTxReceiptLog
+    from rotkehlchen.history.events.structures.evm_event import EvmEvent
+    from rotkehlchen.user_messages import MessagesAggregator
+
+logger = logging.getLogger(__name__)
+log = RotkehlchenLogsAdapter(logger)
+
+
+class Compoundv3CommonDecoder(DecoderInterface):
+
+    def __init__(
+            self,
+            evm_inquirer: 'EvmNodeInquirer',
+            base_tools: 'BaseDecoderTools',
+            msg_aggregator: 'MessagesAggregator',
+            rewards_address: 'ChecksumEvmAddress',
+    ) -> None:
+        super().__init__(
+            evm_inquirer=evm_inquirer,
+            base_tools=base_tools,
+            msg_aggregator=msg_aggregator,
+        )
+        self.rewards_address = rewards_address
+        self.underlying_tokens: dict[EvmToken, EvmToken | None] = {}
```

Comment:
It's the same as previous. Now we don't populate it on initialization, but only cache it when needed. So, if/when a supply/withdraw event is decoded, then a DB query is made and save the result in this cache. This will help if same assets are supplied/withdrawn many times, then it will fetch from the DB only once.

Reply:
as discussed make a note for this to improve. We need a timed cache for this in compound decoder and any other mappings similar to it that keep the cache only for a certain duration. Specifically the decoder run. Timed duration may also make sense .... but maybe at start/end of decoding is cleaner.

Reply:
https://github.com/orgs/rotki/projects/11?pane=issue&itemId=59322844